custom_metrics/third-parties.js

```diff
@@ -0,0 +1,34 @@
+/**
+ * Detect the presence of third party libraries.
+ *
+ * Look for common library aliases on the global scope.
+ * If available, detect libraries' versions.
+ *
+ * Outputs JSON-serialized list of library-version pairs.
+ *     e.g.: `["a@1.0","b","c@2.2"]`
+ */
+var thirdParties = [];
+var name, version;
+
+function addThirdParty(name, version) {
+	if (!name) {
+		return;
+	}
+
+	if (version) {
+		name += '@' + version;
```

Comment:
I would recommend finding a different way to serialise the library name/version pairs. In the wild, there are lots of "unexpected" version strings that will cause issues later. Typically, one would expect something like `1.2.3`, but it is also very common to have `1.3.0-rc.4`, and then (less common but still frequent in large crawls) many constructs such as `1.4.11-local+sha.2c9066e`, `@EDGE` or `@VERSION` or `@@VERSION`, `0.0.0dev (22487)`, `1.9.7build20160503-0010,1.9.7-20130926-IBM (2dfbbaf)`, `0.0.0${project.version} (23843)`, `2.2.3 -ajax/load,-ajax/script,-ajax/jsonp,-deprecated,-effects,-effects/animatedSelector,-effects/Tween,-event/alias,-wrap,-core/ready,-css/hiddenVisibleSelectors` (looks like a custom build of jQuery), `$Rev: 249 $`, and some versions of jQuery Tools return not a string, but a dict/object of all installed plug-ins, which can look like this: `{'tabs': '1.0.1', 'scrollable': '1.0.5', 'expose': '1.0.3', 'tooltip': '1.0.2', 'overlay': '1.0.4'}`.

In our paper, we ignored these special versions because we usually cannot find metadata (such as release dates) for them, and it's very complicated (or impossible) to compare if one of these special versions is older or newer than another. For a general-purpose web crawl, I think it makes sense to keep all versions, but if these special version strings/objects aren't properly "escaped", splitting the string into name/version in a DB query will get quite complicated.